### ISSUE TITLE
Only one update for every client sent

### ISSUE THREAD

**Issue by PatrickJosh** [nextcloud/notify_push#149]
Hello,

I've migrated my Nextcloud instance to another server and since that I'm experiencing problems with notify_push. It seems that every client gets only one update and nothing more. The first update is as fast as expected but if I create/modify/delete files after that, nothing is received.

The contents of the logfile with loglevel `debug`:

<details>
<summary>Log of server</summary>

```
Nov 27 13:14:10 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:10.464091 +00:00] DEBUG [notify_push] src/lib.rs:262: new websocket connection from Some(2003:e0:674f:3c00:6fe1:4b29:6122:2bc9)
Nov 27 13:14:10 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:10.466759 +00:00] DEBUG [notify_push::nc] src/nc.rs:27: Verifying credentials for joshua.noeske
Nov 27 13:14:10 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:10.623239 +00:00] INFO [notify_push::connection] src/connection.rs:66: new websocket authenticated as joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.014794 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files/Test.md
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.014826 +00:00] DEBUG [notify_push::storage_mapping] src/storage_mapping.rs:95: querying storage mapping for 3
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.015344 +00:00] DEBUG [notify_push::connection] src/connection.rs:99: Sending notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.015367 +00:00] DEBUG [notify_push::connection] src/connection.rs:99: Sending notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.015438 +00:00] DEBUG [notify_push::connection] src/connection.rs:99: Sending notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.015446 +00:00] DEBUG [notify_push::connection] src/connection.rs:99: Sending notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.016358 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.016399 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.016408 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.016419 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.016425 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:18 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:18.616685 +00:00] DEBUG [notify_push::connection] src/connection.rs:123: Sending ping to <another user>
Nov 27 13:14:19 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:19.349385 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 1 and path appdata_ochmvfb4enme/text/documents/464624
Nov 27 13:14:19 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:19.349420 +00:00] DEBUG [notify_push::storage_mapping] src/storage_mapping.rs:95: querying storage mapping for 1
Nov 27 13:14:19 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:19.352044 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 1 and path appdata_ochmvfb4enme/text/documents
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.743336 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files/Test.md
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.743386 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.743405 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.743413 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.743420 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.744476 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.744532 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.744551 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.744558 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:29 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:29.744565 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:33 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:33.464381 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 1 and path appdata_ochmvfb4enme/text/documents/464624
Nov 27 13:14:33 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:33.465912 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 1 and path appdata_ochmvfb4enme/text/documents
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.549952 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files_trashbin/files/Test.md.d1638015275
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.550072 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.550104 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.550142 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.550160 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.558659 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.558758 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.558785 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.558821 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.558841 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.569584 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files_trashbin/files
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.569696 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.569732 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.569770 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.569789 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.574197 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files_trashbin
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.574301 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.574334 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.574368 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.574387 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.577595 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.577683 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.577712 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.577748 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.577766 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.580705 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files_trashbin/files
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.580797 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.580826 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.580861 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.580879 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.596183 +00:00] DEBUG [notify_push] src/lib.rs:441: Received storage update notification for storage 3 and path files
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.596280 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.596309 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.596342 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:35 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:35.596361 +00:00] DEBUG [notify_push::connection] src/connection.rs:103: Debouncing notify_file to joshua.noeske
Nov 27 13:14:40 fedora-server-0 notify_push[8096]: [2021-11-27 12:14:40.745464 +00:00] DEBUG [notify_push::connection] src/connection.rs:161: websocket error: WebSocket protocol error: Connection reset without closing handshake
```

</details>

However, the same time in the `test_client`:

<details>
<summary>Log of test_client</summary>

```
Found push server at wss://<url>/push/ws
Authenticated
Received file update notification

```

</details>

The service configuration: 

<details>
<summary>notify_push.service</summary>

```
[Unit]
Description = Push daemon for Nextcloud clients
Documentation=https://github.com/nextcloud/notify_push
After=httpd.service php-fpm.service mariadb.service redis.service

[Service]
Environment = PORT=7867 # Change if you already have something running on this port
Environment=NEXTCLOUD_URL=http://127.0.0.1
ExecStart = /var/www/nextcloud/apps/notify_push/bin/x86_64/notify_push /var/www/nextcloud/config/config.php
User=apache

[Install]
WantedBy = multi-user.target

```

</details>

The relevant parts of my `config.php`:

<details>
<summary>config.php</summary>

```
<?php
$CONFIG = array (
  'trusted_domains' => 
  array (
    0 => '<url>',
    1 => '127.0.0.1',
  ),
  'datadirectory' => '/mnt/usb/nextcloud/data',
  'dbtype' => 'mysql',
  'version' => '22.2.3.0',
  'overwrite.cli.url' => '<url>',
  'dbname' => 'nextcloud',
  'dbhost' => 'localhost:/var/lib/mysql/mysql.sock',
  'dbport' => '',
  'dbtableprefix' => 'oc_',
  'dbuser' => '<user>',
  'dbpassword' => '<password>',
  'installed' => true,
  'memcache.local' => '\\OC\\Memcache\\APCu',
  'memcache.distributed' => '\\OC\\Memcache\\Redis',
  'memcache.locking' => '\\OC\\Memcache\\Redis',
  'redis' => 
  array (
    'host' => '/run/redis/redis-server.sock',
    'port' => '0',
  ),
  'maintenance' => false,
  'loglevel' => 1,
  'twofactor_enforced' => 'true',
  'twofactor_enforced_groups' => 
  array (
  ),
  'twofactor_enforced_excluded_groups' => 
  array (
  ),
  'trusted_proxies' => 
  array (
    0 => '127.0.0.1',
    1 => '::1',
  ),
);

```

</details>

The self test of notify_push shows no errors.

I'm using Fedora 35 with
```
mariadb  Ver 15.1 Distrib 10.5.12-MariaDB, for Linux (x86_64) using  EditLine wrapper
Redis server v=6.2.6 sha=00000000:0 malloc=jemalloc-5.1.0 bits=64 build=db74a1b227e296fe
Server version: Apache/2.4.51 (Fedora)
```
and maybe it's relevant that I'm running with SELinux enforcing.

Before the migration I didn't have any issues, I think, but I'm not completely sure. I can definitely say that I worked some time ago, but I didn't test it extensively since the update to 0.3.0. The server was running Ubuntu 20.04, mariadb 10.3 or 10.2 and also Apache, but I don't know the version right now. But if it is needed, I can look it up.

**Comment by vitobotta** [comment]
Same problem here, and when I use the test_client I get this after the first update:

```
Authenticated
Received file update notification
Error:
   0: WebSocket protocol error: Connection reset without closing handshake
   1: Connection reset without closing handshake

Location:
   /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/result.rs:1915
```

**Comment by gvtulder** [comment]
I have a similar problem. It's not always exactly one update, sometimes there are two or three that reach the test_client before it dies. Sometimes it takes a while for updates to show up: I change a file, it's synced with the server, the notify_push server receives the event, but it doesn't immediately push to the clients.

It might be something introduced in v0.3.0. I downgraded the notify_push binary to that of v0.2.4, and then everything works fine.

**Comment by PatrickJosh** [comment]
The problem still persists with Nextcloud 23.0.0, as I have tested right now.

**Comment by kemoantemo** [comment]
First one is the fastest after that it's hit or miss and when it hits it take about 10 to 15 second to arrive

**Comment by vitobotta** [comment]
Testing right now with 23.0.2, still the same :(

**Comment by alerque** [comment]
I'm not seeing anything like this. Are you by chance running clients behind some sort of proxy?

**Comment by vitobotta** [comment]
> I'm not seeing anything like this. Are you by chance running clients behind some sort of proxy?

Well, behind Caddy. Everything is done with Docker Compose. How else should the client push be configured if not behind a proxy? Thanks!

**Comment by PatrickJosh** [comment]
I guess @alerque meant the clients, so the Nextcloud Desktop programme or the test-client for notify_push. But I am not running them behind a proxy; even when the clients are on the same network as the Nextcloud instance, they have the same problems.

**Comment by vitobotta** [comment]
> I guess @alerque meant the clients, so the Nextcloud Desktop software or the test-client for notify_push. But I am not running them behind a proxy; even when the clients are on the same network as the Nextcloud instance, they have the same problems.

Ah I see. In that case no, the desktop clients are not behind a proxy of any kind.

**Comment by alerque** [comment]
My push notification server is fronted with an Apache proxy. I wonder which server proxy is involved in the problem cases. Perhaps this is a miss-configuration of the server proxy rather than an issue with this daemon.

**Comment by PatrickJosh** [comment]
I just have tested it again, now with Nextcloud 23.0.2. Even if I ensure that all traffic does not leave the server machine, i. e. by using the test-client on the server, configuring notify_push to use 127.0.0.1 as push server and simultaneously blocking all incoming connections on port 443, the error does still occur.
I am still using Apache as web server and proxy for notify_push.

**Comment by PatrickJosh** [comment]
I've also checked the configuration of my web server once again. I am not sure if they are completely correct but at least I get a connection to the push server with them. I need the config for the local access of Nextcloud because I cannot add the IP adress of the server to the trusted proxies as it is changing regularly. Therefore, I configured the push server to communicate via localhost with the Nextcloud, thus my trusted proxies are 127.0.0.1 and ::1. The push server gets the Nextcloud URL http://127.0.0.1 via an environment variable set in the systemd service (see my initial comment).

Here is my Apache configuration:

<details>
<summary>Apache configs</summary>

**Access Nextcloud from localhost:**
```
<VirtualHost 127.0.0.1:80>
    ServerAdmin webmaster@localhost
    DocumentRoot /var/www/nextcloud-subvolume/nextcloud

    <Directory /var/www/nextcloud-subvolume/nextcloud/>
        Require all granted
        Options FollowSymlinks MultiViews
        AllowOverride All

        <IfModule mod_dav.c>
        Dav off
        </IfModule>
    </Directory>
    Protocols h2 h2c http/1.1

    ProxyPass /push/ws ws://127.0.0.1:7867/ws
    ProxyPass /push/ http://127.0.0.1:7867/
    ProxyPassReverse /push/ http://127.0.0.1:7867/
</VirtualHost>
```

**Global nextcloud file:**
```
<IfModule mod_ssl.c>
<VirtualHost *:443>
    ServerName <domain>

    ServerAdmin webmaster@localhost
    DocumentRoot /var/www/nextcloud-subvolume/nextcloud


    <Directory /var/www/nextcloud-subvolume/nextcloud/>
        Require all granted
        AllowOverride All
        Options FollowSymLinks MultiViews

        <IfModule mod_dav.c>
            Dav off
        </IfModule>
    </Directory>

    ProxyPass /push/ws ws://127.0.0.1:7867/ws
    ProxyPass /push/ http://127.0.0.1:7867/
    ProxyPassReverse /push/ http://127.0.0.1:7867/

#    ProxyPass /push/ws  unix:/var/run/notify_push/notify_push.sock|ws://localhost/ws
#    ProxyPass /push/  unix:/var/run/notify_push/notify_push.sock|http://localhost/
#    ProxyPassReverse /push/  unix:/var/run/notify_push/notify_push.sock|http://localhost/

    SSLCertificateFile /etc/letsencrypt/live/<domain>/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/<domain>/privkey.pem
    Include /etc/letsencrypt/options-ssl-apache.conf

    Protocols h2 h2c http/1.1

    <IfModule mod_headers.c>
        Header always set Strict-Transport-Security "max-age=15552000; includeSubDomains"
    </IfModule>

</VirtualHost>

SSLUseStapling On
SSLStaplingCache "shmcb:logs/ssl_stapling(32768)"
SSLSessionCache "shmcb:logs/ssl_session(32768)"

</IfModule>
```

</details>

I tested it without the ProxyPass part in the file for local Nextcloud access as well as with http/2 deactivated.

**Comment by PatrickJosh** [comment]
For me, this issue has been solved with version 0.4.0. Thus, I'll close this.
